brave/src/test/java/brave/features/handler/FinishedChildrenListener.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2013-2020 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.features.handler;
+
+import brave.handler.MutableSpan;
+import brave.handler.SpanListener;
+import brave.internal.weaklockfree.WeakConcurrentMap;
+import brave.propagation.TraceContext;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+import static java.util.Collections.emptyIterator;
+
+public abstract class FinishedChildrenListener extends SpanListener {
+
+  protected abstract void onFinish(MutableSpan parent, Iterator<MutableSpan> children);
+
+  /** This holds the children of the current parent until the former is finished or abandoned. */
+  final WeakConcurrentMap<TraceContext, TraceContext> childToParent = new WeakConcurrentMap<>();
```

Comment:
note this now uses weak maps

Reply:
oops forgot to switch this back to https://github.com/raphw/weak-lock-free as an example shouldn't use our internal types..